cpp/examples/tutorial_examples/dataset_example.cc

```diff
@@ -0,0 +1,177 @@
+  auto schema =
+      arrow::schema({arrow::field("a", arrow::int64()), arrow::field("b", arrow::int64()),
+                     arrow::field("c", arrow::int64())});
+  std::shared_ptr<arrow::Array> array_a;
+  std::shared_ptr<arrow::Array> array_b;
+  std::shared_ptr<arrow::Array> array_c;
+  arrow::NumericBuilder<arrow::Int64Type> builder;
+  ARROW_RETURN_NOT_OK(builder.AppendValues({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}));
+  ARROW_RETURN_NOT_OK(builder.Finish(&array_a));
+  builder.Reset();
+  ARROW_RETURN_NOT_OK(builder.AppendValues({9, 8, 7, 6, 5, 4, 3, 2, 1, 0}));
+  ARROW_RETURN_NOT_OK(builder.Finish(&array_b));
+  builder.Reset();
+  ARROW_RETURN_NOT_OK(builder.AppendValues({1, 2, 1, 2, 1, 2, 1, 2, 1, 2}));
+  ARROW_RETURN_NOT_OK(builder.Finish(&array_c));
+  return arrow::Table::Make(schema, {array_a, array_b, array_c});
+}
+
+// Set up a dataset by writing two Parquet files.
+arrow::Result<std::string> CreateExampleParquetDataset(
+    const std::shared_ptr<arrow::fs::FileSystem>& filesystem,
+    const std::string& root_path) {
+  // Much like CreateTable(), this is utility that gets us the dataset we'll be reading
+  // from. Don't worry, we also write a dataset in the example proper.
+  auto base_path = root_path + "parquet_dataset";
+  ARROW_RETURN_NOT_OK(filesystem->CreateDir(base_path));
+  // Create an Arrow Table
+  ARROW_ASSIGN_OR_RAISE(auto table, CreateTable());
+  // Write it into two Parquet files
+  ARROW_ASSIGN_OR_RAISE(auto output,
+                        filesystem->OpenOutputStream(base_path + "/data1.parquet"));
+  ARROW_RETURN_NOT_OK(parquet::arrow::WriteTable(
+      *table->Slice(0, 5), arrow::default_memory_pool(), output, 2048));
+  ARROW_ASSIGN_OR_RAISE(output,
+                        filesystem->OpenOutputStream(base_path + "/data2.parquet"));
+  ARROW_RETURN_NOT_OK(parquet::arrow::WriteTable(
+      *table->Slice(5), arrow::default_memory_pool(), output, 2048));
+  return base_path;
+}
+
+arrow::Status PrepareEnv() {
+  // Get our environment prepared for reading, by setting up some quick writing.
+  ARROW_ASSIGN_OR_RAISE(auto src_table, CreateTable())
+  std::shared_ptr<arrow::fs::FileSystem> setup_fs;
+  // Note this operates in the directory the executable is built in.
+  char setup_path[256];
+  getcwd(setup_path, 256);
+  ARROW_ASSIGN_OR_RAISE(setup_fs, arrow::fs::FileSystemFromUriOrPath(setup_path));
+  ARROW_ASSIGN_OR_RAISE(auto dset_path, CreateExampleParquetDataset(setup_fs, ""));
+
+  return arrow::Status::OK();
+}
+
+arrow::Status RunMain() {
+  ARROW_RETURN_NOT_OK(PrepareEnv());
+
+  // First, we need a filesystem object, which lets us interact with our local
+  // filesystem starting at a given path. For the sake of simplicity, that'll be
+  // the current directory.
+  std::shared_ptr<arrow::fs::FileSystem> fs;
+  // This feels pretty bad, but I wasn't finding great solutions that're
+  // system-generic -- could use some advice on how to set this up.
```

Comment:
just to make sure - does it not work to use '.'?

Reply:
I would have to double-check, but I know that `FileSystemFromUriOrPath` requires absolute paths, and I do believe that the URI version also didn't do great with '.'

Honestly, the solution was even worse when I wrote this comment, and I may remove the comment.

Reply:
Ah - you could directly create the LocalFileSystem in that case

Reply:
Is that the pattern we want to teach? I thought these creation functions were the way to go.

Reply:
I think you'd have to resort to the individual functions for setting options (which in the case of S3 or something you probably do want). But otherwise then, I'm not sure if there's a great way to do it generically, yeah

Reply:
Leave it as-is, then?

Reply:
I think that's fine unless someone else comes up with something better